### PR TITLE
👁 Fix sight line algorithm 👁

### PR DESCRIPTION
**Pull request by MillhioreBT**
### Pull Request Prelude
- [x] I have followed [proper The Forgotten Server code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
Emulate the behavior with respect to the line of sight of Tibia RL, this topic was already in progress in the following PR #3265, but for some reason the guy who proposed this closed his PR and deleted his repository, I still want to give him credit for inspiring these changes and giving the idea

### Need Testing
I have test this many times, however I do not have a Tibia RL account good enough to run all kinds of tests, it would be good if someone else will test these changes to check that everything is correct

### Any help
Any suggestion or help regarding the current code is welcome, it is possible that much more can be improved

**Issues addressed:** #2891

### GIF's
**Ray Test**
Here I show an example of many rays that verify the accessible places with this algorithm, make your comparisons with Tibia RL
<details>
<summary>Ray animation</summary>

GIF 29-03-2021 07-41-32 p  m
GIF 29-03-2021 07-43-19 p  m

</details>

### CODE REVIEW

**Comment by nekiro** [comment]
The one thing that keeps it (for me atleast) is the code readibilty, not to be rude, but it looks very bad.

**Comment by MillhioreBT** [comment]
> The one thing that keeps it (for me at least) from merging is the code readibiltdy/quality, not to be rude, but it looks very bad.

You are not rude, it is reality, I will give it touch-ups to make it look more beautiful

**Comment by MillhioreBT** [comment]
I have improved the code, now it is more beautiful and has short circuits compared to the previous one, I have also eliminated the use of `std::swap` because it is not necessary, I took some suggestion from @ranisalt, so thanks to he ;)

GIF 30-03-2021 02-32-57 a  m

### Extra Notes
* The magic walls that block projectiles, block the passage as if they were infinitely high walls
* Added a new mechanic that makes it possible to throw things from the top to the surface, as long as there is nothing that blocks the line of sight

**Comment by soul4soul** [comment]
Amazing test scripts!

**Comment by Zbizu** [comment]
I'm not exactly sure if this is intended, but the check for tile flags executes twice in your code

https://user-images.githubusercontent.com/5205079/123412532-f7e77180-d5b1-11eb-872a-a990e9fa8cfa.mp4

**Comment by MillhioreBT** [comment]
https://user-images.githubusercontent.com/28090948/123562544-501e9f00-d77d-11eb-94be-d58a9610d95c.mp4

File: src/map.cpp [hunk 1]
```diff
@@ -529,102 +529,82 @@ bool Map::canThrowObjectTo(const Position& fromPos, const Position& toPos, bool
...
 	}
 
 	uint16_t x0 = fromPos.x, y0 = fromPos.y, x1 = toPos.x, y1 = toPos.y;
-	auto sightClear = checkSightLine(x0, y0, x1, y1, z0);
-	if (sightClear == SIGHTTILESTATE_FIELD || sightClear == SIGHTTILESTATE_BLOCKED && z0 == z1) {
+	if (sameFloor && (isBlockProjectile(x1, y1, z0) || !checkSightLine(x0, y0, x1, y1, z0))) {
 		return false;
 	}
-
+
```
Comment (EPuncker) [inline src/map.cpp:594]: 😛

Comment (MillhioreBT) [reply src/map.cpp]: thanks punk <3,

**Comment by MillhioreBT** [comment]
I have made the last changes, I have tested it and the old mechanics are maintained correctly, unlike now a new algorithm is used equivalent to the one in TibiaRL, I ask you to test the changes.

All this I achieved thanks to @ranisalt, he has guided me throughout the course of these changes <3

**Pull request closed without merging**